pkg/tfbridge/property_path.go

```diff
@@ -112,22 +127,177 @@ func propertyValueTriggersReplacement(
+		subpath resource.PropertyPath, inputsSubVal resource.PropertyValue, entering bool,
+	) (resource.PropertyValue, error) {
+		if !entering {
+			return inputsSubVal, nil
+		}
+		// Do not compare and do not descend into internal properties.
+		{
+			lastPathStep := subpath[len(subpath)-1]
+			if stepStr, ok := lastPathStep.(string); ok {
+				if resource.IsInternalPropertyKey(resource.PropertyKey(stepStr)) {
+					return inputsSubVal, propertyvalue.SkipChildrenError{}
+				}
+			}
+		}
+
+		tfs, _, err := lookupSchemas(propertyPath(subpath), tfs, ps)
+		if err != nil {
+			// TODO log
+			return resource.NewNullProperty(), abortErr
+		}
+
+		relativePath, err := propertyPath(subpath).GetPathRelativeTo(path)
+		if err != nil {
+			return resource.NewNullProperty(), abortErr
+		}
+
+		planSubVal, ok := relativePath.Get(plan)
+		if !ok || planSubVal.IsNull() {
+			return resource.NewNullProperty(), abortErr
+		}
+
+		if tfs.Computed() && inputsSubVal.IsNull() {
+			return planSubVal, nil
+		}
+
+		if tfs.Type() == shim.TypeList || tfs.Type() == shim.TypeMap {
+			// We only need to check the leaf values, so we skip any collection types.
+			return inputsSubVal, nil
+		}
+
+		// We can not descend into nested sets as planning re-orders the elements
+		if tfs.Type() == shim.TypeSet {
+			// TODO: more sophisticated comparison of nested sets
+			if inputsSubVal.DeepEquals(planSubVal) {
+				return inputsSubVal, propertyvalue.SkipChildrenError{}
+			}
+			return resource.NewNullProperty(), abortErr
+		}
 
-	computed := false
-	visitor := func(path walk.SchemaPath, tfs shim.Schema) {
-		if tfs.Computed() {
-			computed = true
+		if inputsSubVal.DeepEquals(planSubVal) {
+			return inputsSubVal, nil
 		}
+
+		return resource.NewNullProperty(), abortErr
+	}
+	_, err := propertyvalue.TransformPropertyValueDirectional(
+		resource.PropertyPath(path),
+		visitor,
+		inputs,
+	)
+	if err == abortErr {
+		return false
+	}
+	contract.AssertNoErrorf(err, "TransformPropertyValue should only return an abort error")
+	return true
+}
+
+// validInputsFromPlan returns true if the given plan property value could originate from the given inputs.
+// Under the hood, it walks the plan and the inputs and checks that all differences stem from computed properties.
+// Any differences coming from properties which are not computed will be rejected.
+// Note that we are relying on the fact that the inputs will have defaults already applied.
+func validInputsFromPlan(
+	path propertyPath,
+	inputs resource.PropertyValue,
+	plan resource.PropertyValue,
+	tfs shim.SchemaMap,
+	ps map[string]*info.Schema,
+) bool {
+	// We walk both the plan and the inputs and check that all differences stem from computed properties.
+	// We first walk over the plan.
+	if !isPlanCompatibleWithInputs(path, inputs, plan, tfs, ps) {
+		return false
 	}
-	walk.VisitSchema(tfs, visitor)
 
-	return computed
+	// We then walk over the inputs.
+	return isInputCompatibleWithPlan(path, plan, inputs, tfs, ps)
```

Comment:
Why does this have to walk twice?

Reply:
We walk it twice because we are recursing over each property value and not over both.

I can rework the recursion to walk both at the same time but I thought this is preferable since it uses existing code.

Reply:
Reworked this with a new walker function which makes this redundant.